### PR TITLE
[SPARK-23413][UI] Fix sorting tasks by Host / Executor ID at the Stage page

### PR DESCRIPTION
**Pull request by attilapiros**
## What changes were proposed in this pull request?

Fixing exception got at sorting tasks by Host / Executor ID:
```
        java.lang.IllegalArgumentException: Invalid sort column: Host
	at org.apache.spark.ui.jobs.ApiHelper$.indexName(StagePage.scala:1017)
	at org.apache.spark.ui.jobs.TaskDataSource.sliceData(StagePage.scala:694)
	at org.apache.spark.ui.PagedDataSource.pageData(PagedTable.scala:61)
	at org.apache.spark.ui.PagedTable$class.table(PagedTable.scala:96)
	at org.apache.spark.ui.jobs.TaskPagedTable.table(StagePage.scala:708)
	at org.apache.spark.ui.jobs.StagePage.liftedTree1$1(StagePage.scala:293)
	at org.apache.spark.ui.jobs.StagePage.render(StagePage.scala:282)
	at org.apache.spark.ui.WebUI$$anonfun$2.apply(WebUI.scala:82)
	at org.apache.spark.ui.WebUI$$anonfun$2.apply(WebUI.scala:82)
	at org.apache.spark.ui.JettyUtils$$anon$3.doGet(JettyUtils.scala:90)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.spark_project.jetty.servlet.ServletHolder.handle(ServletHolder.java:848)
	at org.spark_project.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:584)
```

Moreover some refactoring to avoid similar problems by introducing constants for each header name and reusing them at the identification of the corresponding sorting index.   

## How was this patch tested?

Manually:

screen shot 2018-02-13 at 18 57 10

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala [hunk 1]
```diff
@@ -963,33 +965,60 @@ private[ui] class TaskPagedTable(
...
-    "Shuffle Spill (Memory)" -> TaskIndexNames.MEM_SPILL,
-    "Shuffle Spill (Disk)" -> TaskIndexNames.DISK_SPILL,
-    "Errors" -> TaskIndexNames.ERROR)
+    HEADER_ID -> null.asInstanceOf[String],
+    HEADER_TASK_INDEX -> TaskIndexNames.TASK_INDEX,
+    HEADER_ATTEMPT -> TaskIndexNames.ATTEMPT,
+    HEADER_STATUS -> TaskIndexNames.STATUS,
+    HEADER_LOCALITY -> TaskIndexNames.LOCALITY,
+    HEADER_EXECUTOR -> TaskIndexNames.EXECUTOR,
+    HEADER_HOST -> TaskIndexNames.EXECUTOR,
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: sorting by host and executor is not the same ... you might have executors 1 & 5 on host A, and execs 2,3,4 on host B. 

The 2.2 UI had both executor and host in the same column: https://github.com/apache/spark/blob/branch-2.2/core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala#L1203

I think we either need to go back to having one column for both, or add an index on host.

thoughts @vanzin ?

File: core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala [hunk 2]
```diff
@@ -963,33 +965,60 @@ private[ui] class TaskPagedTable(
...
+  val HEADER_DESER_TIME = "Task Deserialization Time"
+  val HEADER_GC_TIME = "GC Time"
+  val HEADER_SER_TIME = "Result Serialization Time"
+  val HEADER_GETTING_RESULT_TIME = "Getting Result Time"
+  val HEADER_PEAK_MEM = "Peak Execution Memory"
+  val HEADER_ACCUMULATORS = "Accumulators"
+  val HEADER_INPUT_SIZE = "Input Size / Records"
+  val HEADER_OUTPUT_SIZE = "Output Size / Records"
+  val HEADER_SHUFFLE_READ_TIME = "Shuffle Read Blocked Time"
+  val HEADER_SHUFFLE_TOTAL_READS = "Shuffle Read Size / Records"
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: nit: `HEADER_SHUFFLE_TOTAL_READS` -> `HEADER_SHUFFLE_READ_SIZE` ?

File: core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala (hunk 1 above)
Comment (jiangxb1987) [reply core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: Seems we'd better have a new TaskIndexNames for host column.

Comment (vanzin) [reply core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: Hmmm... I agree that the correct thing would be to sort by host and have an index on that. The problem is that we'd be changing the data on disk, breaking compatibility with previous versions of the disk store. So unless that change goes into 2.3.0, that means revving the disk version number, which would require re-parsing all logs. And that kinda sucks. (I hope by the next major version I - or someone - get time to better investigate versioning of the disk data.)

Given this affects 2.3 we could potentially consider it a blocker. @sameeragarwal probably won't be very happy though.

2.2 actually sorts by executor id, and doesn't have a separate host column (added in SPARK-21675). That's one of these small changes I missed while merging all the SHS stuff.

**Comment by vanzin** [comment]
I'm trying to think of a way that we can avoid these issues going forward - not that I expect this code to change much. Maybe have a unit test that makes sure all declared header constants are mapped to some index, or something like that, and fails if you add a new header constant without a mapping.

File: core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala (hunk 1 above)
Comment (squito) [reply core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: another alternative is to disable sorting by host, and just fix sorting by executor.  That could go into 2.3.1 without breaking compatibility.

Comment (attilapiros) [reply core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: I think that is good idea. I can extend taskHeadersAndCssClasses to store Tuple3 objects where the additional Boolean property flags whether the column is sortable. And for a not sortable column we are skipping the headerLink.

Comment (squito) [reply core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: or even go back to the 2.2 behavior, with executor & host in the same column.

I do think having a separate column for host, and having it be sortable, is actually better ... but just trying to think of simple solutions.

Comment (vanzin) [reply core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: Looks like we'll have a new RC, so I'll jump in the bandwagon and mark this one a blocker too. We can then add the new index in 2.3.0.

File: core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala (hunk 2 above)
Comment (attilapiros) [reply core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala]: In the header constants naming I have followed the existing task index names:

```scala   
  HEADER_SHUFFLE_TOTAL_READS -> TaskIndexNames.SHUFFLE_TOTAL_READS,
```

**Comment by attilapiros** [comment]
Yes, of course. The test of @zsxwing is perfect to avoid similar problems in the future.

**Comment by vanzin** [comment]
LGTM pending tests.

**Comment by attilapiros** [comment]
jenkins retest please

**Comment by vanzin** [comment]
ah, flaky tests. retest this please

**Comment by zsxwing** [comment]
LGTM

**Comment by squito** [comment]
Everything that might have changed from this has passed, the failures are known flaky tests:

https://issues.apache.org/jira/browse/SPARK-23369

https://issues.apache.org/jira/browse/SPARK-23390

merging to master / 2.3

**Comment by squito** [comment]
ack I merged to master but screwed up on 2.3 -- fixing that here: https://github.com/apache/spark/pull/20623

**Pull request closed without merging**